sklearn/ensemble/_hist_gradient_boosting/splitting.pyx

```diff
@@ -388,11 +388,16 @@ cdef class Splitter:
                     &left_indices_buffer[offset_in_buffers[thread_idx]],
                     sizeof(unsigned int) * left_counts[thread_idx]
                 )
-                memcpy(
-                    &sample_indices[right_offset[thread_idx]],
-                    &right_indices_buffer[offset_in_buffers[thread_idx]],
-                    sizeof(unsigned int) * right_counts[thread_idx]
-                )
+                if right_counts[thread_idx] > 0:
+                    # If n_threads >= 2, one might have right_counts[-1] = 0 and
+                    # right_offset[-1] = len(sample_indices) leading to evaluating
+                    # &sample_indices[right_offset[-1]] which produces
+                    # IndexError: Out of bounds on buffer access (with boundscheck=True)
```

Comment:
Keeping your whole comment but just adding a caveat about this being only a rightmost node issue


```suggestion
                    # If we're splitting the rightmost node of the tree,
                    # i.e. the rightmost node in the partition array,
                    # and if n_threads >= 2, one might have right_counts[-1] = 0 and
                    # right_offset[-1] = len(sample_indices) leading to evaluating
                    # &sample_indices[right_offset[-1]] = &samples_indices[n_samples_at_node] = &partition[n_samples_in_tree] which produces
                    # IndexError: Out of bounds on buffer access (with boundscheck=True)
```